# Extending Development Patterns with Tails
## Getting Started
This tutorial focuses on extending the developent patterns beyond the tail. 

Be sure to make sure your packages are updated. For more info on how to update your pakages, visit [Keeping Packages Updated](https://chainladder-python.readthedocs.io/en/latest/install.html#keeping-packages-updated).

In [1]:
# Black linter, optional
%load_ext lab_black

import pandas as pd
import numpy as np
import chainladder as cl
import matplotlib.pyplot as plt

print("pandas: " + pd.__version__)
print("numpy: " + np.__version__)
print("chainladder: " + cl.__version__)

pandas: 1.3.4
numpy: 1.20.3
chainladder: 0.8.9


## Disclaimer
Note that a lot of the examples shown might not be applicable in a real world scenario, and is only meant to demonstrate some of the functionalities included in the package. The user should always follow all applicable laws, the Code of Professional Conduct, applicable Actuarial Standards of Practice, and exercise their best actuarial judgement.

## Basic Tail Fitting

Tails are another class of transformers. Similar to the `Development` estimator, they come with `fit`, `transform` and `fit_transform` methods. Also, like our `Development` estimator, you can define a tail in the absence of data or if you believe development will continue beyond your latest evaluation period.

In [2]:
quarterly = cl.load_sample("quarterly")
quarterly["paid"]

3     6     9      12     15     18     21     24      27      30      33      36      39      42      45      48      51      54      57      60      63      66      69      72      75      78      81      84      87      90      93      96      99      102     105     108     111     114     117     120     123     126     129     132     135
1995  3.0  24.0  65.0  141.0  273.0  418.0  550.0  692.0   814.0   876.0   916.0   959.0   968.0  1002.0  1020.0  1031.0  1041.0  1035.0  1045.0  1054.0  1060.0  1068.0  1075.0  1079.0  1081.0  1084.0  1086.0  1089.0  1091.0  1094.0  1095.0  1096.0  1097.0  1098.0  1098.0  1099.0  1099.0  1100.0  1098.0  1098.0  1098.0  1099.0  1099.0  1100.0  1100.0
1996  1.0  16.0  54.0  135.0  260.0  398.0  594.0  758.0   871.0   964.0  1017.0  1052.0  1089.0  1108.0  1141.0  1175.0  1194.0  1208.0  1217.0  1226.0  1231.0  1243.0  1249.0  1252.0  1253.0  1256.0  1258.0  1261.0  1266.0  1267.0  1268.0  1288.0  1288.0  1289.0  1291.0  1296.0  1296.0  1297.0  1298.0  1298.0  1298.0     NaN     NaN     NaN     NaN
1997  1.0  17.0  55.0  166.0  296.0  442.0  587.0  701.0   811.0   891.0   940.0   976.0  1000.0  1022.0  1043.0  1059.0  1068.0  1080.0  1091.0  1099.0  1104.0  1130.0  1135.0  1138.0  1141.0  1147.0  1163.0  1184.0  1185.0  1186.0  1189.0  1192.0  1194.0  1195.0  1197.0  1197.0  1198.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN
1998  1.0  11.0  40.0   93.0  185.0  343.0  474.0  643.0   744.0   831.0   902.0   951.0   989.0  1022.0  1053.0  1074.0  1093.0  1132.0  1145.0  1162.0  1177.0  1192.0  1197.0  1213.0  1225.0  1237.0  1265.0  1274.0  1275.0  1278.0  1286.0  1288.0  1293.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN
1999  1.0  14.0  47.0  113.0  225.0  379.0  570.0  715.0   832.0   955.0  1048.0  1118.0  1183.0  1214.0  1293.0  1327.0  1363.0  1383.0  1396.0  1438.0  1457.0  1474.0  1504.0  1521.0  1532.0  1547.0  1559.0  1565.0  1573.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN
2000  1.0   6.0  28.0  100.0  194.0  297.0  415.0  521.0   616.0   697.0   758.0   810.0   859.0   892.0   916.0   935.0   950.0   967.0   982.0  1004.0  1013.0  1028.0  1036.0  1046.0  1054.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN
2001  1.0   7.0  37.0  128.0  271.0  427.0  579.0  722.0   838.0   937.0  1020.0  1091.0  1160.0  1207.0  1239.0  1305.0  1323.0  1347.0  1365.0  1375.0  1387.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN
2002  1.0  10.0  45.0  110.0  236.0  442.0  668.0  890.0  1078.0  1198.0  1325.0  1412.0  1524.0  1615.0  1653.0  1720.0  1760.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN
2003  1.0   9.0  31.0   94.0  192.0  299.0  408.0  792.0   873.0   949.0  1019.0  1064.0  1100.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN
2004  4.0  16.0  49.0  170.0  289.0  442.0  601.0  793.0   948.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN
2005  1.0   7.0  36.0   97.0  183.0    NaN    NaN    NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

Upon fitting data, we get updated `ldf_` and `cdf_` attributes that extend beyond the length of the triangle. Notice how the tail includes extra development periods (age 147) beyond the end of the triangle (age 135) at which point an age-to-ultimate tail factor is applied.

In [3]:
tail = cl.TailCurve()
tail.fit(quarterly)

print("Triangle latest", quarterly.development.max())
tail.fit(quarterly).ldf_["paid"]

Triangle latest 135


3-6       6-9      9-12     12-15     15-18     18-21     21-24     24-27     27-30     30-33     33-36     36-39    39-42    42-45     45-48     48-51    51-54     54-57     57-60     60-63     63-66     66-69     69-72     72-75     75-78     78-81     81-84     84-87     87-90     90-93     93-96     96-99    99-102   102-105   105-108   108-111   111-114   114-117  117-120  120-123   123-126  126-129  129-132  132-135   135-138   138-141   141-144   144-147  147-150
(All)  8.5625  3.554745  2.765914  1.933185  1.605535  1.401081  1.327029  1.165767  1.109803  1.077971  1.054556  1.046539  1.03534  1.03039  1.028639  1.017245  1.01494  1.010918  1.014197  1.008495  1.013206  1.008549  1.007365  1.005104  1.006258  1.009568  1.006634  1.002668  1.001661  1.002694  1.005374  1.001645  1.000838  1.001117  1.001673  1.000278  1.000835  0.999583      1.0      1.0  1.000911      1.0  1.00091      1.0  1.000119  1.000097  1.000079  1.000065  1.00029

These extra twelve months (147 - 135, or one year) of development patterns are included as it is typical to want to track IBNR run-off over a 1-year time horizon from the valuation date. The one-year extension is currently fixed at one year and there is no ability to extend it even further. However, a subsequent version of `chainladder` will look to address this issue.  

## Curve Fitting

Curve fitting takes selected development patterns and extrapolates them using either an `exponential` or `inverse_power` fit.  In most cases, the `inverse_power` produces a thicker (more conservative) tail.

In [4]:
exp = cl.TailCurve(curve="exponential").fit(quarterly["paid"])
exp.tail_

135-Ult
(All)  1.00065

In [5]:
inv = cl.TailCurve(curve="inverse_power").fit(quarterly["paid"])
inv.tail_

135-Ult
(All)  1.021283

When fitting a tail, by default, all of the data will be used; however, we can specify which period of development patterns we want to begin including in the curve fitting process with `fit_period`. 

Patterns will also be generated for 100 periods beyond the end of the triangle by default, or we can specify how far beyond the triangle to project the tail factor to before dropping the age-to-age factor down to 1.0 using `extrap_periods`.

Note that even though we can extrapolate the curve many years beyond the end of the triangle for computational purposes, the resultant development factors will compress all `ldf_` beyond one year into a single age-ultimate factor.

In [6]:
quarterly["incurred"]

3      6      9      12     15     18      21      24      27      30      33      36      39      42      45      48      51      54      57      60      63      66      69      72      75      78      81      84      87      90      93      96      99      102     105     108     111     114     117     120     123     126     129     132     135
1995  44.0   96.0  194.0  420.0  621.0  715.0   748.0   906.0   950.0   973.0   997.0  1030.0  1020.0  1035.0  1055.0  1072.0  1070.0  1051.0  1062.0  1070.0  1069.0  1076.0  1081.0  1088.0  1089.0  1091.0  1090.0  1094.0  1094.0  1095.0  1098.0  1096.0  1097.0  1097.0  1101.0  1098.0  1099.0  1103.0  1100.0  1098.0  1100.0  1100.0  1098.0  1101.0  1100.0
1996  42.0  136.0  202.0  365.0  541.0  651.0   817.0   988.0  1052.0  1122.0  1139.0  1173.0  1169.0  1174.0  1205.0  1225.0  1238.0  1228.0  1239.0  1254.0  1249.0  1262.0  1264.0  1267.0  1266.0  1263.0  1278.0  1273.0  1269.0  1279.0  1281.0  1299.0  1296.0  1302.0  1303.0  1300.0  1300.0  1302.0  1300.0  1303.0  1300.0     NaN     NaN     NaN     NaN
1997  17.0   43.0  135.0  380.0  530.0  714.0   813.0   945.0   966.0  1008.0  1028.0  1069.0  1064.0  1073.0  1086.0  1112.0  1100.0  1111.0  1115.0  1128.0  1128.0  1142.0  1170.0  1159.0  1155.0  1167.0  1177.0  1194.0  1196.0  1198.0  1197.0  1198.0  1201.0  1201.0  1200.0  1203.0  1200.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN
1998  10.0   43.0  107.0  238.0  393.0  574.0   732.0   894.0   935.0   967.0  1019.0  1037.0  1062.0  1079.0  1099.0  1138.0  1126.0  1163.0  1198.0  1207.0  1209.0  1215.0  1216.0  1240.0  1243.0  1259.0  1281.0  1284.0  1286.0  1289.0  1292.0  1297.0  1298.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN
1999  13.0   41.0  109.0  306.0  481.0  657.0   821.0  1007.0  1021.0  1141.0  1171.0  1249.0  1267.0  1289.0  1358.0  1400.0  1400.0  1409.0  1437.0  1468.0  1476.0  1488.0  1524.0  1550.0  1550.0  1561.0  1566.0  1585.0  1583.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN
2000   2.0   29.0   88.0  254.0  380.0  501.0   615.0   735.0   788.0   842.0   912.0   915.0   953.0   963.0   977.0   990.0  1001.0  1005.0  1013.0  1029.0  1030.0  1056.0  1056.0  1066.0  1066.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN
2001   4.0   25.0  151.0  333.0  777.0  663.0   856.0   988.0  1063.0  1167.0  1199.0  1242.0  1307.0  1305.0  1348.0  1362.0  1362.0  1376.0  1376.0  1383.0  1411.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN
2002   2.0   34.0  115.0  290.0  472.0  809.0  1054.0  1543.0  1617.0  1505.0  1599.0  1695.0  1818.0  1716.0  1819.0  1839.0  1820.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN
2003   3.0   19.0   90.0  692.0  597.0  929.0   883.0  1117.0  1092.0  1176.0  1198.0  1230.0  1221.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN
2004   4.0   38.0  138.0  371.0  583.0  756.0   902.0  1111.0  1212.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN
2005  21.0   79.0  115.0  299.0  422.0    NaN     NaN     NaN     NaN

In [7]:
cl.TailCurve(fit_period=(12, None), extrap_periods=50).fit(quarterly).ldf_["incurred"]

3-6       6-9      9-12     12-15     15-18     18-21    21-24     24-27     27-30     30-33     33-36    36-39     39-42     42-45     45-48     48-51     51-54     54-57    57-60     60-63     63-66     66-69    69-72     72-75     75-78     78-81     81-84     84-87     87-90    90-93     93-96     96-99    99-102   102-105   105-108   108-111   111-114   114-117   117-120   120-123  123-126   126-129   129-132   132-135   135-138   138-141   141-144   144-147   147-150
(All)  3.598765  2.476844  2.734072  1.468338  1.296558  1.182523  1.24184  1.045144  1.043969  1.036461  1.036835  1.02265  0.997308  1.032489  1.019202  0.997929  1.005544  1.011627  1.01173  1.003865  1.010892  1.009946  1.00807  0.999864  1.006029  1.008043  1.005945  0.999689  1.003302  1.00144  1.004519  1.000409  1.001669  1.001111  0.999168  0.999445  1.002501  0.997921  1.000417  0.999584      1.0  0.998182  1.002732  0.999092  1.000298  1.000257  1.000221  1.000191  1.001191

In [8]:
cl.TailCurve(fit_period=(1, None), extrap_periods=50).fit(quarterly).ldf_["incurred"]

3-6       6-9      9-12     12-15     15-18     18-21    21-24     24-27     27-30     30-33     33-36    36-39     39-42     42-45     45-48     48-51     51-54     54-57    57-60     60-63     63-66     66-69    69-72     72-75     75-78     78-81     81-84     84-87     87-90    90-93     93-96     96-99    99-102   102-105   105-108   108-111   111-114   114-117   117-120   120-123  123-126   126-129   129-132   132-135   135-138   138-141   141-144   144-147   147-150
(All)  3.598765  2.476844  2.734072  1.468338  1.296558  1.182523  1.24184  1.045144  1.043969  1.036461  1.036835  1.02265  0.997308  1.032489  1.019202  0.997929  1.005544  1.011627  1.01173  1.003865  1.010892  1.009946  1.00807  0.999864  1.006029  1.008043  1.005945  0.999689  1.003302  1.00144  1.004519  1.000409  1.001669  1.001111  0.999168  0.999445  1.002501  0.997921  1.000417  0.999584      1.0  0.998182  1.002732  0.999092  1.000184  1.000154  1.000129  1.000108  1.000565

In this example, we ignore the first five development patterns for curve fitting, and we allow our tail extrapolation to go 50 quarters beyond the end of the triangle.  Note that both `fit_period` and `extrap_periods` follow the `development_grain` of the triangle being fit.

## Chaining Multiple Transformers


It is very common to need to get the development factors first then apply a tail curve to extend our development pattern. `chainladder` transformers take `Triangle` objects as inputs, but the returned objects are also `Triangle` objects with their `transform` method. To chain multiple transformers together, we must invoke the `transform` method on each transformer similar to how `sklearn` approaches its own tranformers.   

In [9]:
print("First attempt:")
try:
    cl.TailCurve().fit(cl.Development().fit(quarterly))
    print("This works.")
except:
    print("This fails because we did not transform the triangle")

print("\nSecond attempt:")
try:
    cl.TailCurve().fit(cl.Development().fit_transform(quarterly))
    print("This passes because we transformed the triangle")
except:
    print("This fails.")

First attempt:
This fails because we did not transform the triangle

Second attempt:
This passes because we transformed the triangle


We can also invoke the methods without chaining the operations together.

In [10]:
dev = cl.Development().fit_transform(quarterly)
tail = cl.TailCurve().fit(dev)
tail.cdf_["paid"]

3-Ult       6-Ult      9-Ult     12-Ult    15-Ult    18-Ult    21-Ult    24-Ult    27-Ult    30-Ult    33-Ult    36-Ult    39-Ult    42-Ult   45-Ult    48-Ult    51-Ult    54-Ult   57-Ult    60-Ult    63-Ult    66-Ult    69-Ult    72-Ult    75-Ult    78-Ult  81-Ult    84-Ult   87-Ult    90-Ult   93-Ult   96-Ult    99-Ult   102-Ult   105-Ult  108-Ult   111-Ult   114-Ult   117-Ult   120-Ult   123-Ult  126-Ult  129-Ult  132-Ult  135-Ult   138-Ult   141-Ult   144-Ult  147-Ult
(All)  946.336189  110.521015  31.091127  11.240816  5.814662  3.621635  2.584887  1.947875  1.670896  1.505578  1.396678  1.324423  1.265527  1.222331  1.18628  1.153252  1.133702  1.117013  1.10495  1.089482  1.080305  1.066224  1.057186  1.049456  1.044127  1.037633  1.0278  1.021026  1.01831  1.016621  1.01389  1.00847  1.006814  1.005971  1.004849  1.00317  1.002891  1.002054  1.002472  1.002472  1.002472  1.00156  1.00156  1.00065  1.00065  1.000531  1.000434  1.000355  1.00029

Chaining multiple transformers together is a very common pattern in `chainladder`.  Like its inspiration `sklearn`, we can create an overall estimator known as a `Pipeline` that combines multiple transformers and optional predictors in one estimator. 

In [11]:
sequence = [
    ("simple_dev", cl.Development(average="simple")),
    ("inverse_power_tail", cl.TailCurve(curve="inverse_power")),
]

pipe = cl.Pipeline(steps=sequence).fit(quarterly)

`Pipeline` keeps references to each step with its `named_steps` argument.

In [12]:
print(pipe.named_steps.simple_dev)
print(pipe.named_steps.inverse_power_tail)

Development(average='simple')
TailCurve(curve='inverse_power')


The `Pipeline` estimator is almost an exact replica of the `sklearn Pipeline`, and the docs for `sklearn` are very comprehensive. To learn more about `Pipeline`, [reference their docs](https://scikit-learn.org/stable/modules/compose.html#pipeline).

With a `Triangle` transformed to include development patterns and tails, we are now ready to start fitting our suite of IBNR models.